script/prerender/contentPages.tsx

```diff
@@ -0,0 +1,223 @@
+      loader: bookLoader,
+    };
+  }));
+};
+
+export type Pages = Array<{code: number, page: AnyMatch}>;
+
+export const prepareErrorPages = (): Promise<Pages> => Promise.resolve([
+  {code: 404, page: {route: notFound}},
+]);
+
+export const prepareDeveloperPages = (): Promise<Pages> => Promise.resolve(
+  process.env.REACT_APP_ENV === 'development'
+    ? [{code: 200, page: {route: developerHome}}]
+    : []
+);
+
+export const prepareBookPages = (
+  bookLoader: ReturnType<AppServices['archiveLoader']['book']>,
+  book: Book
+) => asyncPool(20, findTreePages(book.tree), (section) =>
+  prepareContentPage(bookLoader, book.slug, stripIdVersion(section.id))
+    .then((page) => ({code: 200, page}))
+);
+
+export const getBookSitemap = (
+  bookLoader: ReturnType<AppServices['archiveLoader']['book']>,
+  pages: Array<{page: Match<typeof content>}>
+) => pages.map((record) => {
+  const matchState = assertDefined(
+    record.page.state,
+    'match state wasn\'t defined, it should have been'
+  );
+  const archivePage = assertDefined(
+    bookLoader.page(matchState.pageUid).cached(),
+    'page wasn\'t cached, it should have been'
+  );
+
+  return {
+    changefreq: EnumChangefreq.MONTHLY,
+    lastmod: moment(archivePage.revised).format('YYYY-MM-DD'),
+    url: matchUrl(record.page),
+  };
+});
+
+type RenderPages = (
+  services: AppOptions['services'],
+  pages: Pages
+) => Promise<void>;
+export const renderPages: RenderPages = async(services, pages) => {
+  const renderPage = makeRenderPage(services);
+  await asyncPool(50, pages, ({code, page}) => renderPage(page, code));
+};
+
+interface Options {
+  body: string;
+  styles: ServerStyleSheet;
+  fonts: FontCollector['fonts'];
+  meta: Meta[];
+  state: AppState;
+  modules: string[];
+  title: string;
+}
+function injectHTML(html: string, {body, styles, state, fonts, meta, modules, title}: Options) {
+
+  const assetManifest = JSON.parse(readAssetFile('asset-manifest.json'));
+
+  const extractAssets = () => Object.keys(assetManifest)
+    .filter((asset) => modules.indexOf(asset.replace('.js', '')) > -1)
+    .map((k) => assetManifest[k]);
+
+  const scripts = extractAssets().map(
+    (c) => `<script type="text/javascript" src="${c}"></script>`
+  );
+
+  html = html.replace(/<title>.*?<\/title>/, `<title>${title}</title>`);
+
+  html = html.replace('</head>',
+    fonts.map((font) => `<link rel="stylesheet" href="${font}">`).join('') +
+    meta.map(
+      (tag) => `<meta ${Object.entries(tag).map(([name, value]) => `${name}="${value}"`).join(' ')} />`).join(''
+    ) +
+    styles.getStyleTags() +
+    '</head>'
+  );
+  html = html.replace(
+    '<div id="root"></div>',
+    `<div id="root">${body}</div>` +
+    `<script>window.__PRELOADED_STATE__ = ${JSON.stringify(state).replace(/</g, '\\u003c')}</script>`
+  );
+
+  html = html.replace(/(<script src="[^"]+\.chunk\.js"><\/script>)/, scripts.join('') + '$1');
+
+  return html;
+}
```

Comment:
We've chatted about this before, but I'm super curious how well the regex based html manipulation will work.  I've always heard "never use regexes on HTML".  If it does prove buggy there's several DOM parsers that could be used but they're probably slower than regex.

Reply:
yea using regex on html is... limited, only useful for simple replacements, nothing that involves any structural knowledge of the document. for now all the regex manipulations are simpler, or at least easier, than booting a dom parser, processing the whole document, and doing some traversal/manipulations on it.

if i ever start to feel like i'm putting a square peg in a round hole i'll definitely think about switching to using dom traversal for the manipulations.